lib/BuildSystem/BuildSystem.cpp

```diff
@@ -302,14 +306,24 @@ class BuildSystemImpl : public BuildSystemCommandInterface {
     buildWasAborted = value;
   }
 
+  void resetForBuild() {
+    std::lock_guard<std::mutex> guard(stateMutex);
+    isCancelled_ = false;
+  }
+
   /// Cancel the running build.
   void cancel() {
+    std::lock_guard<std::mutex> guard(stateMutex);
+
     isCancelled_ = true;
-    getExecutionQueue().cancelAllJobs();
+    // Cancel jobs if we actually have a queue.
+    if (executionQueue.get() != nullptr)
+      getExecutionQueue().cancelAllJobs();
   }
 
   /// Check if the build has been cancelled.
   bool isCancelled() {
+    std::lock_guard<std::mutex> guard(stateMutex);
     return isCancelled_;
```

Comment:
It would be nice to avoid this lock, since we check this in many places inside the system. I think we can safely do so by making isCancelled_ atomic, we only actually need the lock for the ordering w.r.t. the executionQueue.